clang/lib/CodeGen/CGExpr.cpp

```diff
@@ -1113,8 +1114,9 @@ llvm::Value *CodeGenFunction::GetCountedByFieldExprGEP(
   llvm::Value *Res = nullptr;
   if (const auto *DRE = dyn_cast<DeclRefExpr>(StructBase)) {
     Res = EmitDeclRefLValue(DRE).getPointer(*this);
-    Res = Builder.CreateAlignedLoad(ConvertType(DRE->getType()), Res,
-                                    getPointerAlign(), "dre.load");
+    if (isArrow)
+      Res = Builder.CreateAlignedLoad(ConvertType(DRE->getType()), Res,
+                                      getPointerAlign(), "dre.load");
```

Comment:
But wait, now I’m a bit confused, shouldn’t we already be performing `LValueToRValue` conversion if the LHS of `->` is an lvalue?

Reply:
Ah, I see, I think `StructAccessBase` just unwraps everything and hands you the DRE, stripping the lvalue to rvalue conversion in the process.

Since this code was originally for handling `__builtin_dynamic_object_size`, does that even support or handle cases like `s.x` correctly?

Reply:
> I see, so if I’m reading this correctly, this is trying to deal with the fact that in a.b, a is an lvalue of some struct type (say S), but in a->b, a is an lvalue to a S*, so we need to load it to get a prvalue S*, which is functionally identical to an lvalue S in that both are just a pointer in IR.

Um...I think so? I've always thought of an l-value as an r-value, but I suppose that's not correct anymore with the advent of [a-z]r-values in C++. I'm sure there's a better way to do what I need this to do, but I haven't found the front-end approved way of doing it.

Reply:
This confusion is why I've repeatedly requested in all the related reviews that we stop ignoring casts.

Reply:
> This confusion is why I've repeatedly requested in all the related reviews that we stop ignoring casts.

Yeah, that makes sense, but doesn’t the implementation of `__builtin_dynamic_object_size` do that too (or is that what you’re referring to?)

Reply:
> This confusion is why I've repeatedly requested in all the related reviews that we stop ignoring casts.

Then how do I handle degenerative cases like `&*&((struct s *)(char *)&ptr)->array[123]`? There doesn't seem to be an easy way to get down to what the actual type `ptr` is unless we strip away every single cast and unary op. Or am I missing something entirely? I haven't seen a solution in all of related reviews...

Reply:
> Then how do I handle degenerative cases like `&*&((struct s *)(char *)&ptr)->array[123]`?

At a glance, *imo*, we should just not support those cases, but if support for them is really necessary, then I’m not entirely sure either. Maybe just do the unwrapping to get to the type, but then evaluate the original LHS of the MemberExpr? I haven’t thought too hard about that though, so there might be problems w/ that too...

Reply:
And even if I can get the base type of `ptr`, I still need to query whether or not `ptr` has the `counted_by` attribute and then build a reference to the `count` object from the original `Expr`.

Reply:
> > Then how do I handle degenerative cases like `&*&((struct s *)(char *)&ptr)->array[123]`?
> 
> At a glance, _imo_, we should just not support those cases, but if support for them is really necessary, then I’m not entirely sure either. Maybe just do the unwrapping to get to the type, but then evaluate the original LHS of the MemberExpr? I haven’t thought too hard about that though, so there might be problems w/ that too...

I don't like this option. It's too restrictive, because C is what it is and it allows for such things. And I don't want to run into a situation where some program (e.g. Linux) has macros that expand into something horrible and Clang just throws up its hands and gives up.

This cannot be the ONLY situation where this issue has come up...

Reply:
In this context, you should never need to look through an lvalue-to-rvalue conversion, I think?  The counted_by attribute has to refer to a field at a fixed offset from the array, so the operand of an lvalue-to-rvalue conversion will never contain the record you're looking for.

If you want to specifically look through noop casts, or array->pointer decay casts, that's fine.  (All casts have a CastKind you can use to distinguish.)

Reply:
The different casts are annoying and unsolved as of now, but they aren't even the issue I'm currently trying to solve.

In the case that @kees put in [this comment](https://github.com/llvm/llvm-project/pull/102549#issuecomment-2291865451) has this for the second if-then statement:

```
if (__builtin_get_counted_by(stack.array)) {
  ...
}

|   |-IfStmt 0x556b13fd5840 <line:18:9, line:20:9>
|   | |-CallExpr 0x556b13fd5710 <line:18:13, col:49> 'unsigned char *'
|   | | |-ImplicitCastExpr 0x556b13fd56f8 <col:13> 'int (*)()' <BuiltinFnToFnPtr>
|   | | | `-DeclRefExpr 0x556b13fd5688 <col:13> '<builtin fn type>' Function 0x556b13fd52a0 '__builtin_get_counted_by' 'int ()'
|   | | `-MemberExpr 0x556b13fd56c8 <col:38, col:44> 'short[] __counted_by(count)':'short[]' lvalue .array 0x556b13fcefc8
|   | |   `-DeclRefExpr 0x556b13fd56a8 <col:38> 'struct counted' lvalue Var 0x556b13fd51f0 'stack' 'struct counted'
```

There's no implicit casts of `stack.array` or even a conversion from l-value to r-value cast. Okay fine. However, there's a segfault when trying to emit the conditional:

```
clang: /home/morbo/llvm/llvm-project/clang/lib/CodeGen/CGExprScalar.cpp:960: Value *(anonymous namespace)::ScalarExprEmitter::EmitConversionToBool(Value *, QualType): Assertion `(SrcType->isIntegerType() || isa<llvm::PointerType>(Src->getType())) && "Unknown scalar type to convert"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: ./bin/clang -O2 /tmp/x.c -o - -S -emit-llvm
1.	<eof> parser at end of file
2.	/tmp/x.c:9:5: LLVM IR generation of declaration 'main'
3.	/tmp/x.c:9:5: Generating code for declaration 'main'
 #0 0x000055d6e5b500f8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (./bin/clang+0x7e240f8)
 #1 0x000055d6e5b4dcae llvm::sys::RunSignalHandlers() (./bin/clang+0x7e21cae)
 #2 0x000055d6e5ab6ef6 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x00007f3afac591a0 (/lib/x86_64-linux-gnu/libc.so.6+0x3d1a0)
 #4 0x00007f3afaca70ec __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007f3afac59102 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #6 0x00007f3afac424f2 abort ./stdlib/abort.c:81:7
 #7 0x00007f3afac42415 _nl_load_domain ./intl/loadmsgcat.c:1177:9
 #8 0x00007f3afac51d32 (/lib/x86_64-linux-gnu/libc.so.6+0x35d32)
 #9 0x000055d6e5f17598 (anonymous namespace)::ScalarExprEmitter::EmitScalarConversion(llvm::Value*, clang::QualType, clang::QualType, clang::SourceLocation, (anonymous namespace)::ScalarExprEmitter::ScalarConversionOpts) CGExprScalar.cpp:0:0
#10 0x000055d6e5f163f3 clang::CodeGen::CodeGenFunction::EmitScalarConversion(llvm::Value*, clang::QualType, clang::QualType, clang::SourceLocation) (./bin/clang+0x81ea3f3)
#11 0x000055d6e5eea4f8 clang::CodeGen::CodeGenFunction::EvaluateExprAsBool(clang::Expr const*) (./bin/clang+0x81be4f8)
#12 0x000055d6e5ec05c1 clang::CodeGen::CodeGenFunction::EmitBranchOnBoolExpr(clang::Expr const*, llvm::BasicBlock*, llvm::BasicBlock*, unsigned long, clang::Stmt::Likelihood, clang::Expr const*) (./bin/clang+0x81945c1)
#13 0x000055d6e5ed4143 clang::CodeGen::CodeGenFunction::EmitIfStmt(clang::IfStmt const&) (./bin/clang+0x81a8143)
```

This isn't an issue the base `DeclRefExpr` has a pointer type. It seems that the GEP is getting the wrong struct type or something. The "type" that's bad here is a non-pointer to the structure type.

So okay. It'd probably be best to generate an `Expr` that refers to the `count` variable and take it's address, right? That seems iffy to do during CodeGen, and I shouldn't have to, given that I have all of the data I need readily available. But of course the LLVM IR has no "get me the address of this object" instruction, so I have to generate the GEP. Okay fine, but the call to `EmitDeclRefLValue` apparently doesn't do the Right Thing(tm) so that I can use that to refer to the `count` variable. The GEP that's generated looks fine, but there's still the segfault. That's why I added the `isArrow` check. I no longer think that's correct, but what else have we?

Reply:
> The "type" that's bad here is a non-pointer to the structure type.

Yeah, that sounds like you’re getting an lvalue of the structure’s type instead of an rvalue of the corresponding pointer type (which is the same in the IR, but codegen really doesn’t like type mismatches there in my experience).

Seems like you need to somehow turn the lvalue into a pointer (if that’s possible in codegen; maybe you’re supposed to generate an AST node for it). I’m not really familiar enough w/ codegen to know how you would go about this. Maybe `LValue::getPointer()` is what you need, but that’s just me guessing here...

Reply:
I can't see how you could crash like that; the only thing I can imagine is that EmitBuiltinExpr is somehow returning something that isn't a pointer, but in that case you've seriously messed up somewhere.

I guess the new approach sort of works?  It doesn't really solve the issue that you're not computing the base correctly, though.

Please add report_fatal_error messages to codegen if the expression doesn't have the expected form, instead of silently returning null.

Reply:
It was a mystery to me too. I'm sure I messed something up...

I'm happier with the new approach (generating a new `MemberExpr`). I don't have to duplicate code that already generates LLVM IR for `Exprs`. 

For this builtin, I took @Sirraide's advice and limited the kinds of the arguments we accept to a few "sensible" forms, namely `ptr->array` and `&ptr->array[0]`. It appears to work pretty well, given the limitations we have on the `counted_by` argument (the 'count' needs to be within the same non-anonymous struct as the FAM). For completeness, I made sure that `MemberExprs` were generated for anonymous structs / unions that may contain the `count`, and the new `MemberExpr` starts from the proper base `Expr`.

I'll add a report_fatal_error message.